### ISSUE TITLE
Bug in format/write

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6624]
Migrated from [rt.perl.org#22977](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22977) (status was 'resolved')

Searchable as RT22977$

**Comment by p5pRT** [comment]
### From schopper@usm.uni-muenchen.de

Hi there\,

I belive\, that we've found a bug in perl\. We discovered it when changing from
perl\-5\.4\.4 to perl\-5\.8\.0\, where we got strange results with
write in a program we heavily used without problems\.

We found\, that write no longer use the actual values of the variables
but get stuck with those during the first call to write\. This behaviour depends
on wether the writing subroutine ist declared or not\.

A sample program would be&#8203;:
==========================
&nbsp;  1  \#\!/usr/lib/perl5/neu/usr\-bin/perl \-w
&nbsp;  2
&nbsp;  3  use strict;
&nbsp;  4
&nbsp;  5  sub show\_woche \(\); \# Comment out to get right result\!
&nbsp;  6
&nbsp;  7  sub show\_woche \(\)
&nbsp;  8  \{ my $tag;
&nbsp;  9    my $eintrag;
 10    my $datum;
 11
 12    my @&#8203;tag = \(
 13     \[0 \,     \[\['a'\]\, \['b'\]\]\]\,
 14     \[50000 \, \[\['d'\]\, \['e'\]\]\]\,
 15    \);
 16
 17    foreach $tag \(@&#8203;tag\)
 18    \{ my $eintrag\_liste = $tag\->\[1\];
 19      $datum = $tag\->\[0\];
 20      foreach $eintrag \(@&#8203;\{$eintrag\_liste\}\)
 21      \{ my @&#8203;test = \($datum\, $eintrag\->\[0\]\);
 22        \#printf "%\-8s  %\-10s\\n"\, @&#8203;test;
 23        write;
 24
 25        format STDOUT =
 26  @&#8203;\<\<\<\<\<\<\<  @&#8203;\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<
 27  @&#8203;test
 28  \.
 29      \}
 30    \}
 31  \}
 32
 33  show\_woche;
 34
 35  exit 0;

This gives the wrong output&#8203;:
0         a
0         a
0         a
0         a

But when we comment out line 5 we get the right result&#8203;:
0         a
0         b
50000     d
50000     e



The old \(correct\) version of perl is&#8203;:
=====================================
Summary of my perl5 \(5\.0 patchlevel 4 subversion 4\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.0\.30\, archname=i586\-linux
&nbsp;    uname='linux lar 2\.0\.30 \#1 mon jun 9 14&#8203;:40&#8203;:34 mesz 1997 i686 '
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    bincompat3=n useperlio=undef d\_sfio=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, optimize='\-O2 \-fomit\-frame\-pointer \-pipe \-m486 \-malign\-loops=2 \-malign\-jumps=2 \-malign\-functions=2'\, gccversion=2\.7\.2\.2\.f\.2
&nbsp;    cppflags='\-Dbool=char \-DHAS\_BOOL \-I/usr/local/include'
&nbsp;    ccflags ='\-Dbool=char \-DHAS\_BOOL \-I/usr/local/include'
&nbsp;    stdchar='char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    voidflags=15\, castflags=0\, d\_casti32=define\, d\_castneg=define
&nbsp;    intsize=4\, alignbytes=4\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags ='\-L/usr/local/lib \-s'
&nbsp;    libpth=/usr/local/lib /shlib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-ldbm \-ldb \-ldl \-lm \-lc
&nbsp;    libc=/lib/libc\.so\.5\.4\.33\, so=so
&nbsp;    useshrplib=true\, libperl=libperl\.so
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic \-Wl\,\-rpath\,/usr/lib/perl5/i586\-linux/5\.00404/CORE'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Built under linux
&nbsp;  Compiled at Dec  4 1997 17&#8203;:44&#8203;:17
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/lib/perl5/i586\-linux/5\.00404
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/lib/perl5/site\_perl/i586\-linux
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    /usr/lib/perl5/i586\-linux
&nbsp;    \.


The new \(buggy\) Version is&#8203;:
===========================
Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.21\-pre5\, archname=i686\-linux
&nbsp;    uname='linux lar 2\.4\.21\-pre5 \#1 smp thu feb 27 12&#8203;:10&#8203;:15 mez 2003 i686 '
&nbsp;    config\_args=''
&nbsp;    hint=previous\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-fomit\-frame\-pointer \-mcpu=athlon\-mp \-falign\-loops=1 \-falign\-jumps=1 \-falign\-functions=1 \-mpreferred\-stack\-boundary=2'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    ccversion=''\, gccversion='3\.2\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags ='\-s'
&nbsp;    libpth=/usr/i686\-pc\-linux\-gnulibc2/lib /usr/i686\-pc\-linux\-gnulibc2/local/lib
&nbsp;    libs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.3\.2'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Built under linux
&nbsp;  Compiled at Mar 25 2003 10&#8203;:46&#8203;:58
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/lib/perl5/5\.8\.0/i686\-linux
&nbsp;    /usr/lib/perl5/5\.8\.0
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.0/i686\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.0
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    \.


Thanks in advance for any help you can give\.

With best regards


Rüdiger Schopper


GPG/PGP\-Key&#8203;: 
wwwkeys\.pgp\.net&#8203;:11371/pks/lookup?op=vindex&search=0x97A5CFE9&fingerprint=on
Fingerprint = B44F 7109 A40F B383 AA1B 778D 43B9 767C 97A5 CFE9

\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/                                                                    \_/
\_/            \(\`\-'\-\,\-\. \_\.                                             \_/
\_/                  '\.\_ \(\_ '\.\_\_\.\-\,                                    \_/
\_/                  \.\-;\`\-\.\-\`\_\.\.\-'   \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/                  \`\\\_\_\_\-\.\(\_       \_/                                \_/
\_/                     \_\.\)  a'\-\.\_   \_/  Rüdiger Schopper              \_/
\_/                \_\_ \.'          \`\\ \_/  schopper@&#8203;usm\.uni\-muenchen\.de  \_/
\_/              \.'  \`          \.\_^/ \_/                                \_/
\_/      \_\.\.\.\_\.\-'          /\`\\ /     \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/   \_\.'              \\  /   \)      \_/                                \_/
\_/  \(/      /      \)   |/           \_/  The surest sign that intelli\- \_/
\_/   |      |      /   ;\-\.          \_/  gent life in universe exists  \_/
\_/   ;\\     /'\-\.\_\_\.|  /\-\. \\         \_/  is that none of it has tried  \_/
\_/   / \)   /       | |   '\.\\        \_/  to contact us \!               \_/
\_/  / /|  /        \\ |    ||        \_/                                \_/
\_/ | / / /         | |   / |        \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/ ||  \\ \\         | |   \\\_/                                          \_/
\_/ \)\_\`\\ \)\_\`\\       \)\_\`\\                                               \_/
\_/ \`""\` \`""\`       \`""\`                                               \_/
\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\*\*\* This message was entirely written with recycled electrons \*\*\*

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Jul 14\, 2003 at 05&#8203;:37&#8203;:42AM \-0000\, Ruediger Schopper wrote&#8203;:
> I belive\, that we've found a bug in perl\. We discovered it when changing from
> perl\-5\.4\.4 to perl\-5\.8\.0\, where we got strange results with
> write in a program we heavily used without problems\.

The bug can be reduced to the simpler case&#8203;:

&nbsp;    sub f \($\); \# Comment out to get right result\!
&nbsp;    sub f \($\) \{
&nbsp;    my $test = $\_\[0\];
&nbsp;    write;
&nbsp;    format STDOUT =
&nbsp;    @&#8203;\<\<\<\<\<\<\<
&nbsp;    $test
&nbsp;    \.
&nbsp;    \}
&nbsp;    f\(1\);
&nbsp;    f\(2\);

It's due to there being multiple CVs associated with f due to to the
forward declaration\, and to newATTRSUB\(\) copying the contents of the
second CV to the first\.  This causes any nested subs to have CvOUTSIDE
pointing to the wrong CV\. A call to pad\_fixup\_inner\_anons\(\) fixes up any
nested anon subs\, but nested formats aren't fixed up\.

Since I'm currently working on trying to get sub declarations
to take effect at the start of a sub rather than at the end \(eg
sub f\($\) \{ f\(1\) \}\)\, I'll see if I can work in a fix for this at the
same time\.

\-\- 
"You're so sadly neglected\, and often ignored\.
A poor second to Belgium\, When going abroad\."
Monty Python \- "Finland"

**Comment by p5pRT** [comment]
### From schopper@usm.uni-muenchen.de

Hi Dave\,


thanks for the extremely fast response\. I would like to know if we might be
able to help you with the subroutine/format bug? We are clearly not deep
enough into the stuff to be able to effectively offer time\, but perhaps we
might help with money\. Do you think we can be of any help\, or shall we just
wait until you come up with a bugfix? 

Our problem is\, that we have to upgrade perl eventually\, so we are in principle
perfectly willing to contibute in order to get over that bug\.

I don't want to annoy you\, so just give us a short notice\, if you like\.

Best regards

Rüdiger

> On Mon\, Jul 14\, 2003 at 05&#8203;:37&#8203;:42AM \-0000\, Ruediger Schopper wrote&#8203;:
> > I belive\, that we've found a bug in perl\. We discovered it when changing from
> > perl\-5\.4\.4 to perl\-5\.8\.0\, where we got strange results with
> > write in a program we heavily used without problems\.
> 
> The bug can be reduced to the simpler case&#8203;:
> 
>     sub f \($\); \# Comment out to get right result\!
>     sub f \($\) \{
>     my $test = $\_\[0\];
>     write;
>     format STDOUT =
>     @&#8203;\<\<\<\<\<\<\<
>     $test
>     \.
>     \}
>     f\(1\);
>     f\(2\);
> 
> It's due to there being multiple CVs associated with f due to to the
> forward declaration\, and to newATTRSUB\(\) copying the contents of the
> second CV to the first\.  This causes any nested subs to have CvOUTSIDE
> pointing to the wrong CV\. A call to pad\_fixup\_inner\_anons\(\) fixes up any
> nested anon subs\, but nested formats aren't fixed up\.
> 
> Since I'm currently working on trying to get sub declarations
> to take effect at the start of a sub rather than at the end \(eg
> sub f\($\) \{ f\(1\) \}\)\, I'll see if I can work in a fix for this at the
> same time\.
> 
> \-\- 
> "You're so sadly neglected\, and often ignored\.
> A poor second to Belgium\, When going abroad\."
> Monty Python \- "Finland"

GPG/PGP\-Key&#8203;: 
wwwkeys\.pgp\.net&#8203;:11371/pks/lookup?op=vindex&search=0x97A5CFE9&fingerprint=on
Fingerprint = B44F 7109 A40F B383 AA1B 778D 43B9 767C 97A5 CFE9

\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/                                                                    \_/
\_/            \(\`\-'\-\,\-\. \_\.                                             \_/
\_/                  '\.\_ \(\_ '\.\_\_\.\-\,                                    \_/
\_/                  \.\-;\`\-\.\-\`\_\.\.\-'   \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/                  \`\\\_\_\_\-\.\(\_       \_/                                \_/
\_/                     \_\.\)  a'\-\.\_   \_/  Rüdiger Schopper              \_/
\_/                \_\_ \.'          \`\\ \_/  schopper@&#8203;usm\.uni\-muenchen\.de  \_/
\_/              \.'  \`          \.\_^/ \_/                                \_/
\_/      \_\.\.\.\_\.\-'          /\`\\ /     \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/   \_\.'              \\  /   \)      \_/                                \_/
\_/  \(/      /      \)   |/           \_/  The surest sign that intelli\- \_/
\_/   |      |      /   ;\-\.          \_/  gent life in universe exists  \_/
\_/   ;\\     /'\-\.\_\_\.|  /\-\. \\         \_/  is that none of it has tried  \_/
\_/   / \)   /       | |   '\.\\        \_/  to contact us \!               \_/
\_/  / /|  /        \\ |    ||        \_/                                \_/
\_/ | / / /         | |   / |        \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/ ||  \\ \\         | |   \\\_/                                          \_/
\_/ \)\_\`\\ \)\_\`\\       \)\_\`\\                                               \_/
\_/ \`""\` \`""\`       \`""\`                                               \_/
\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\*\*\* This message was entirely written with recycled electrons \*\*\*

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Jul 21\, 2003 at 05&#8203;:43&#8203;:39PM \+0200\, Ruediger Schopper wrote&#8203;:
> Hi Dave\,
> thanks for the extremely fast response\. I would like to know if we might be
> able to help you with the subroutine/format bug? We are clearly not deep
> enough into the stuff to be able to effectively offer time\, but perhaps we
> might help with money\. Do you think we can be of any help\, or shall we just
> wait until you come up with a bugfix? 

Thanks for the kind offer\, but I don't think that giving money to me
personally would speed up the bugfix\.

However\, if you want to contribute financially to the development of Perl
generally\, you might consider a donation to the Perl Foundation\,
http&#8203;://perlfoundation\.org/\, which \(amongst other things\) provides grants
to various individuals associated with Perl\.

I am hoping to have a patch ready in 2\-3 weeks time\, but it's
unlikely to be included in the soon\-to\-be\-released 5\.8\.1\, so it will most
likely appear in 5\.10\.0 \(or possibly a 5\.8\.2\)\. Either of these may be a
year or more away\.

Of course\, I may find that I can't fix it after all\.\.\.

Regards\,

Dave\.


\-\- 
"But Sidley Park is already a picture\, and a most amiable picture too\.
The slopes are green and gentle\. The trees are companionably grouped at
intervals that show them to advantage\. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged\." Lady Croom \- Arcadia

**Comment by p5pRT** [comment]
### From schopper@usm.uni-muenchen.de

Hi Dave\,

do you have any news concerning the subroutine/format bug I sent you a
couple of weeks ago? If so\, we would be happy hear from you\.

With best regards

Rüdiger

> I am hoping to have a patch ready in 2\-3 weeks time\, but it's
> unlikely to be included in the soon\-to\-be\-released 5\.8\.1\, so it will most
> likely appear in 5\.10\.0 \(or possibly a 5\.8\.2\)\. Either of these may be a
> year or more away\.
> 
> Of course\, I may find that I can't fix it after all\.\.\.
> 
> Regards\,
> 
> Dave\.
> 
> 
> \-\- 
> "But Sidley Park is already a picture\, and a most amiable picture too\.
> The slopes are green and gentle\. The trees are companionably grouped at
> intervals that show them to advantage\. The rill is a serpentine ribbon
> unwound from the lake peaceably contained by meadows on which the right
> amount of sheep are tastefully arranged\." Lady Croom \- Arcadia

GPG/PGP\-Key&#8203;: 
wwwkeys\.pgp\.net&#8203;:11371/pks/lookup?op=vindex&search=0x97A5CFE9&fingerprint=on
Fingerprint = B44F 7109 A40F B383 AA1B 778D 43B9 767C 97A5 CFE9

\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/                                                                    \_/
\_/            \(\`\-'\-\,\-\. \_\.                                             \_/
\_/                  '\.\_ \(\_ '\.\_\_\.\-\,                                    \_/
\_/                  \.\-;\`\-\.\-\`\_\.\.\-'   \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/                  \`\\\_\_\_\-\.\(\_       \_/                                \_/
\_/                     \_\.\)  a'\-\.\_   \_/  Rüdiger Schopper              \_/
\_/                \_\_ \.'          \`\\ \_/  schopper@&#8203;usm\.uni\-muenchen\.de  \_/
\_/              \.'  \`          \.\_^/ \_/                                \_/
\_/      \_\.\.\.\_\.\-'          /\`\\ /     \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/   \_\.'              \\  /   \)      \_/                                \_/
\_/  \(/      /      \)   |/           \_/  The surest sign that intelli\- \_/
\_/   |      |      /   ;\-\.          \_/  gent life in universe exists  \_/
\_/   ;\\     /'\-\.\_\_\.|  /\-\. \\         \_/  is that none of it has tried  \_/
\_/   / \)   /       | |   '\.\\        \_/  to contact us \!               \_/
\_/  / /|  /        \\ |    ||        \_/                                \_/
\_/ | / / /         | |   / |        \_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\_/ ||  \\ \\         | |   \\\_/                                          \_/
\_/ \)\_\`\\ \)\_\`\\       \)\_\`\\                                               \_/
\_/ \`""\` \`""\`       \`""\`                                               \_/
\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/\_/
\*\*\* This message was entirely written with recycled electrons \*\*\*

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Sep 02\, 2003 at 01&#8203;:28&#8203;:58PM \+0200\, Ruediger Schopper wrote&#8203;:
> Hi Dave\,
> 
> do you have any news concerning the subroutine/format bug I sent you a
> couple of weeks ago? If so\, we would be happy hear from you\.

I'm working on a opath\, bug have got side\-tracked with things more
directlky related to the shortly\-to\-be\-released 5\.8\.1\. It also turned out
to be more compilated than I had thought\. So no ETA\, I'm afraid :\-\(

> > I am hoping to have a patch ready in 2\-3 weeks time\, but it's
> > unlikely to be included in the soon\-to\-be\-released 5\.8\.1\, so it will most
> > likely appear in 5\.10\.0 \(or possibly a 5\.8\.2\)\. Either of these may be a
> > year or more away\.
> > 
> > Of course\, I may find that I can't fix it after all\.\.\.

\-\- 
"Strange women lying in ponds distributing swords is no basis for a system
of government\. Supreme executive power derives from a mandate from the
masses\, not from some farcical aquatic ceremony\."
Dennis \- Monty Python and the Holy Grail\.

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Sep 02\, 2003 at 09&#8203;:01&#8203;:05PM \+0100\, Dave Mitchell wrote&#8203;:

Would have helped if I'd proof\-read it before hitting the send key\.
Here it is again in English&#8203;:

I'm working on a patch\, but have got side\-tracked with things more
directly related to the shortly\-to\-be\-released 5\.8\.1\. It also turned out
to be more complicated than I had thought\. So no ETA\, I'm afraid :\-\(

\-\- 
To collect all the latest movies\, simply place an unprotected ftp server
on the Internet\, and wait for the disk to fill\.\.\.\.

**Comment by p5pRT** [comment]
### From Alexander.stohr@gmx.de

It seems to me that this bug is still a pending one\.
The below sample code still raises the problem\.
\-Alex\.

\#\!/usr/bin/perl \-w
use strict;
sub show \(\);
my $set\_glob;
show;
sub show \(\)
\{ my @&#8203;tag\_list = \(\[100\]\, \[101\]\, \[102\]\);
&nbsp;  my $set\_sub = \[99\];
&nbsp;  foreach $set\_sub \(@&#8203;tag\_list\)
&nbsp;  \{ $set\_glob = $set\_sub;
&nbsp;    print "$set\_sub  $set\_sub\->\[0\]         \(loop\)\\n";
&nbsp;    $~ = 'TEST\_SUB';
&nbsp;    write;
&nbsp;    $~ = 'TEST\_GLOB';
&nbsp;    write;
&nbsp;    print "\\n";
\# this is the buggy part
\# it happens only for variables defined in subroutines
\# the bug origin might be the perl interpreter
\# the bug is present in at least&#8203;: perl\, v5\.8\.2 built for i686\-linux
format TEST\_SUB =
@&#8203;\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<  @&#8203;\<\<\<\<\<\<\<\<\<  \(sub\)
\{ $set\_sub\, $set\_sub\->\[0\]
\}       
\.       
&nbsp;  \}     
\}       
format TEST\_GLOB =
@&#8203;\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<  @&#8203;\<\<\<\<\<\<\<\<\<  \(glob\)
\{ $set\_glob\, $set\_glob\->\[0\]
\}       
\.       

Results are&#8203;:
ARRAY\(0x81241c8\)  100         \(loop\)
ARRAY\(0x81242e8\)  99          \(sub\)
ARRAY\(0x81241c8\)  100         \(glob\)

ARRAY\(0x81317a4\)  101         \(loop\)
ARRAY\(0x81242e8\)  99          \(sub\)
ARRAY\(0x81317a4\)  101         \(glob\)

ARRAY\(0x8131888\)  102         \(loop\)
ARRAY\(0x81242e8\)  99          \(sub\)
ARRAY\(0x8131888\)  102         \(glob\)

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Mon Jul 14 15&#8203;:15&#8203;:03 2003\, davem wrote&#8203;:
> On Mon\, Jul 14\, 2003 at 05&#8203;:37&#8203;:42AM \-0000\, Ruediger Schopper wrote&#8203;:
> > I belive\, that we've found a bug in perl\. We discovered it when
> changing from
> > perl\-5\.4\.4 to perl\-5\.8\.0\, where we got strange results with
> > write in a program we heavily used without problems\.
> 
> The bug can be reduced to the simpler case&#8203;:
> 
>     sub f \($\); \# Comment out to get right result\!
>     sub f \($\) \{
>     my $test = $\_\[0\];
>     write;
>     format STDOUT =
>     @&#8203;\<\<\<\<\<\<\<
>     $test
>     \.
>     \}
>     f\(1\);
>     f\(2\);
> 
> It's due to there being multiple CVs associated with f due to to the
> forward declaration\, and to newATTRSUB\(\) copying the contents of the
> second CV to the first\.  This causes any nested subs to have 
CvOUTSIDE
> pointing to the wrong CV\. A call to pad\_fixup\_inner\_anons\(\) fixes up
> any
> nested anon subs\, but nested formats aren't fixed up\.
> 
> Since I'm currently working on trying to get sub declarations
> to take effect at the start of a sub rather than at the end \(eg
> sub f\($\) \{ f\(1\) \}\)\, I'll see if I can work in a fix for this at the
> same time\.
> 

It appears that things got worse for 5\.10\.0/blead\.\.\.

perl\-5\.8\.8 rt\-22977\.pl
1
1

\(Instead of 1 2\)

perl\-5\.9\.0 rt\-22977\.pl
Segmentation fault

perl\-5\.10\.0 rt\-22977\.pl
Segmentation fault

perl\-blead rt\-22977\.pl
Segmentation fault

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Fri\, 27 Jun 2008 12&#8203;:04&#8203;:09 \-0700\, "Bram via RT" \<perlbug\-followup@&#8203;perl\.org> said&#8203;:

&nbsp;  > It appears that things got worse for 5\.10\.0/blead\.\.\.

Binary search reveals\.\.\.

\-\-\-\-Program\-\-\-\-
sub f \($\); \# Comment out to get right result\!
sub f \($\) \{
&nbsp;        my $test = $\_\[0\];
&nbsp;        write;
&nbsp;        format STDOUT =
@&#8203;\<\<\<\<\<\<\<
$test
\.
\}
f\(1\);
f\(2\);

\-\-\-\-Output of \.\.\./pMZAcVt/perl\-5\.8\.0@&#8203;19648/bin/perl\-\-\-\-
1
1

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\./pgL2gwc/perl\-5\.8\.0@&#8203;19649/bin/perl\-\-\-\-

\-\-\-\-EOF \($?='11'\)\-\-\-\-


Change 19649 by rgs@&#8203;rgs\-home on 2003/05/31 18&#8203;:33&#8203;:07

&nbsp;    Subject&#8203;: \[PATCH\] jumbo closure patch broke formats
&nbsp;    From&#8203;: Dave Mitchell \<davem@&#8203;fdgroup\.com>
&nbsp;    Date&#8203;: Sat\, 31 May 2003 19&#8203;:54&#8203;:48 \+0100
&nbsp;    Message\-ID&#8203;: \<20030531185448\.GA6055@&#8203;fdgroup\.com>
&nbsp;    Plus restore the original test script for bug \#22372



\-\- 
andreas

**Comment by p5pRT** [comment]
### From @nwc10

Dave notes&#8203;:

Was a format/closure/newATTSUB bug in 5\.8\.8; in 5\.10\.0\, maint\, blead\, it
gets worse and becomes an assertion failure

**Comment by p5pRT** [comment]
### From chromatic@wgz.org

On Friday 27 June 2008 12&#8203;:04&#8203;:09 Bram via RT wrote&#8203;:

> > The bug can be reduced to the simpler case&#8203;:
> >
> >     sub f \($\); \# Comment out to get right result\!
> >     sub f \($\) \{
> >     my $test = $\_\[0\];
> >     write;
> >     format STDOUT =
> >     @&#8203;\<\<\<\<\<\<\<
> >     $test
> >     \.
> >     \}
> >     f\(1\);
> >     f\(2\);

That's the same bug as RT \#50528\.

\-\- c

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Attached patch fixes bug \#22977\.  make regen\_perly after applying\.

\-zefram

**Comment by p5pRT** [comment]
### From zefram@fysh.org

<details><summary>Inline Patch</summary>

```diff
diff --git a/perly.y b/perly.y
index 18e5875..a61a6b3 100644
--- a/perly.y
+++ b/perly.y
@@ -511,7 +511,9 @@ peg	:	PEG
 	;
 
 format	:	FORMAT startformsub formname block
-			{ SvREFCNT_inc_simple_void(PL_compcv);
+			{
+			  CV *fmtcv = PL_compcv;
+			  SvREFCNT_inc_simple_void(PL_compcv);
 #ifdef MAD
 			  $$ = newFORM($2, $3, $4);
 			  prepend_madprops($1->tk_mad, $$, 'F');
@@ -521,6 +523,10 @@ format	:	FORMAT startformsub formname block
 			  newFORM($2, $3, $4);
 			  $$ = (OP*)NULL;
 #endif
+			  if (CvOUTSIDE(fmtcv) && !CvUNIQUE(CvOUTSIDE(fmtcv))) {
+			    SvREFCNT_inc_simple_void(fmtcv);
+			    pad_add_anon((SV*)fmtcv, OP_NULL);
+			  }
 			}
 	;
 
diff --git a/t/comp/form_scope.t b/t/comp/form_scope.t
new file mode 100644
index 0000000..3ef891e
--- /dev/null
+++ b/t/comp/form_scope.t
@@ -0,0 +1,18 @@
+#!./perl
+#
+# Tests bug #22977.  Test case from Dave Mitchell.
+
+print "1..2\n";
+
+sub f ($);
+sub f ($) {
+my $test = $_[0];
+write;
+format STDOUT =
+ok @<<<<<<<
+$test
+.
+}
+
+f(1);
+f(2);
```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

2009/12/15 Zefram \<zefram@&#8203;fysh\.org>&#8203;:
> Attached patch fixes bug \#22977\.  make regen\_perly after applying\.

Thanks\, applied\, regenerated\, and MANIFEST completed\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Dec 15 02&#8203;:49&#8203;:25 2009\, rgs@&#8203;consttype\.org wrote&#8203;:
> 2009/12/15 Zefram \<zefram@&#8203;fysh\.org>&#8203;:
> > Attached patch fixes bug \#22977\. �make regen\_perly after applying\.
> 
> Thanks\, applied\, regenerated\, and MANIFEST completed\.

The commit id was 421f30ed1e9\.

The commit message did not explain&#8203;:  It follows Dave Mitchell’s
suggestion of adding formats to '&' entries in the enclosing sub’s pad\,
so that the fixup code for subs applies to formats as well\.

\-\- 

Father Chrysostomos
